controllers/hcpvaultsecretsapp_controller.go

```diff
@@ -319,3 +340,93 @@ func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// dynamic)
+	secretsListParams := &hvsclient.ListAppSecretsParams{
+		Context: ctx,
+		AppName: appName,
+		// Type is currently non-functional, so we have to filter the list
+		// ourselves
+		// Type: ptr.To(helpers.HVSSecretTypeDynamic),
+	}
+	appSecretsList, err := c.ListAppSecrets(secretsListParams, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list app Secrets for app %s: %w", appName, err)
+	}
+
+	dynamicSecrets := make([]*models.Secrets20231128OpenSecret, 0)
+
+	if appSecretsList.Payload != nil {
+		// TODO(tvoran): only fetch/create dynamic secrets that are new or are
+		// due for rotation
+		for _, appSecret := range appSecretsList.Payload.Secrets {
+			if appSecret.Type != helpers.HVSSecretTypeDynamic {
+				continue
+			}
+			secretParams := &hvsclient.OpenAppSecretParams{
+				Context:    ctx,
+				AppName:    appName,
+				SecretName: appSecret.Name,
+			}
+			dynamicResp, err := c.OpenAppSecret(secretParams, nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get dynamic secret %s in app %s: %w",
+					appSecret.Name, appName, err)
+			}
+			if dynamicResp != nil && dynamicResp.Payload != nil {
+				dynamicSecrets = append(dynamicSecrets, dynamicResp.Payload.Secret)
+			}
+		}
+	}
+
+	return dynamicSecrets, nil
+}
+
+// getDynamicRenewPercent returns the default renewal percent or the synconfig
+// dynamic renewal percent in that order of precendence
+func getDynamicRenewPercent(syncConfig *secretsv1beta1.HVSSyncConfig) int {
+	renewPercent := defaultDyanmicRenewPercent
+	if syncConfig != nil && syncConfig.Dynamic != nil && syncConfig.Dynamic.RenewalPercent != 0 {
+		renewPercent = syncConfig.Dynamic.RenewalPercent
+	}
+	return renewPercent
+}
+
+func makeHVSDynamicStatus(secret *models.Secrets20231128OpenSecret) secretsv1beta1.HVSDynamicStatus {
+	status := secretsv1beta1.HVSDynamicStatus{
+		Name: secret.Name,
+	}
+	if secret.DynamicInstance != nil {
+		status.CreatedAt = secret.DynamicInstance.CreatedAt.String()
+		status.ExpiresAt = secret.DynamicInstance.ExpiresAt.String()
+		status.TTL = secret.DynamicInstance.TTL
+	}
+	return status
+}
+
+// getNextRequeue returns whichever is less between the current `requeueAfter`
+// and the next renewal time of the dynamic secret instance
+func getNextRequeue(requeueAfter time.Duration, dynamicInstance *models.Secrets20231128OpenSecretDynamicInstance, renewPercent int, now time.Time) time.Duration {
+	if dynamicInstance == nil {
+		return requeueAfter
+	}
+	nextRequeue := requeueAfter
+
+	// Calculate the time when the dynamic secret should be renewed
+	fullTTL := time.Time(dynamicInstance.ExpiresAt).Sub(time.Time(dynamicInstance.CreatedAt))
+	renewPoint := fullTTL * time.Duration(renewPercent) / 100
+	renewTime := time.Time(dynamicInstance.CreatedAt).Add(renewPoint)
+	howLongUntilRenewTime := renewTime.Sub(now)
+
+	if howLongUntilRenewTime < requeueAfter {
+		nextRequeue = howLongUntilRenewTime
+	}
+	if nextRequeue < 0 {
+		nextRequeue = 1 * time.Second
```

Comment:
Could we have a constant `defaultNextRequeue` for 1s?

Reply:
Sure, added one in fd0656f